File: benchs/fp_triad_compare.c
#include <stdio.h>
#include <stdlib.h>

static double triad_sum(const double *b, const double *c, const double *d, int n, double a) {
    double acc = 0.0;
    for (int i = 0; i < n; i++) {
        acc += b[i] + a * c[i] + d[i];
    }
    return acc;
}

int main(void) {
    const int n = 2000;
    const int rounds = 200;
    const double a = 0.31415926535;

    double *b = (double *)malloc((size_t)n * sizeof(double));
    double *c = (double *)malloc((size_t)n * sizeof(double));
    double *d = (double *)malloc((size_t)n * sizeof(double));
    if (!b || !c || !d) {
        free(b);
        free(c);
        free(d);
        return 1;
    }

    for (int i = 0; i < n; i++) {
        b[i] = (double)i * 0.5 + 1.1;
        c[i] = (double)i * 0.25 + 2.3;
        d[i] = (double)i * 0.125 + 3.7;
    }

    double total = 0.0;
    for (int r = 0; r < rounds; r++) {
        total += triad_sum(b, c, d, n, a);
    }

    printf("%.12f\n", total);
    free(b);
    free(c);
    free(d);
    return 0;
}

File: benchmarks/bench/fp_triad_compare.py
#!/usr/bin/env python3


def make_b(n: int):
    return [i * 0.5 + 1.1 for i in range(n)]


def make_c(n: int):
    return [i * 0.25 + 2.3 for i in range(n)]


def make_d(n: int):
    return [i * 0.125 + 3.7 for i in range(n)]


def triad_sum(b, c, d, a: float) -> float:
    acc = 0.0
    for bi, ci, di in zip(b, c, d):
        acc += bi + a * ci + di
    return acc


if __name__ == "__main__":
    n = 2000
    rounds = 200
    a = 0.31415926535
    b = make_b(n)
    c = make_c(n)
    d = make_d(n)
    total = 0.0
    for _ in range(rounds):
        total += triad_sum(b, c, d, a)
    print(f"{total:.12f}")
